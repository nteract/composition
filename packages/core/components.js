// @flow
module.exports = require("./").components;
