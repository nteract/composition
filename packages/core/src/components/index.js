import Notebook from "./notebook";

const _nextgen = require("./ng");

export { _nextgen, Notebook };
