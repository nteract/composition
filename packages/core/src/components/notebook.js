/* eslint-disable no-return-assign */
/* @flow */
import * as React from "react";
import { DragDropContext as dragDropContext } from "react-dnd";
import HTML5Backend from "react-dnd-html5-backend";
import { connect } from "react-redux";
import { List as ImmutableList, Map as ImmutableMap } from "immutable";

import { displayOrder, transforms } from "@nteract/transforms";

import Cell from "../components/cell/cell";

import DraggableCell from "./draggable-cell";
import CellCreator from "../providers/cell-creator";
import StatusBar from "./status-bar";

import {
  focusNextCell,
  focusNextCellEditor,
  moveCell,
  focusCell,
  executeCell
} from "../actions";

import { LinkExternalOcticon } from "@nteract/octicons";

import { light, dark } from "../../themes";

// NOTE: PropTypes are required for the sake of contextTypes
const PropTypes = require("prop-types");

type Props = {
  displayOrder: Array<string>,
  // TODO: Make types stricter, we have definitions _somewhere_
  cellOrder: ImmutableList<any>,
  cellMap: ImmutableMap<string, any>,
  transforms: Object,
  cellPagers: ImmutableMap<string, any>,
  stickyCells: ImmutableMap<string, any>,
  transient: ImmutableMap<string, any>,
  cellFocused: string,
  editorFocused: string,
  theme: string,
  lastSaved: Date,
  kernelSpecDisplayName: string,
  executionState: string,
  models: ImmutableMap<string, any>,
  language: string
};

export class StickyCellContainer extends React.Component<*, *> {
  stickyCellsPlaceholder: ?HTMLElement;
  stickyCellContainer: ?HTMLElement;

  componentDidUpdate(prevProps: Props): void {
    if (this.stickyCellsPlaceholder && this.stickyCellContainer) {
      // Make sure the document is vertically shifted so the top non-stickied
      // cell is always visible.
      this.stickyCellsPlaceholder.style.height = `${
        this.stickyCellContainer.clientHeight
      }px`;
    }
  }

  render() {
    if (
      !this.props.children ||
      React.Children.count(this.props.children) === 0
    ) {
      return null;
    }

    return (
      <div>
        <div
          className="sticky-cells-placeholder"
          ref={ref => {
            this.stickyCellsPlaceholder = ref;
          }}
        />
        <div
          className="sticky-cell-container"
          ref={ref => {
            this.stickyCellContainer = ref;
          }}
        >
          {this.props.children}
        </div>
        <style jsx>{`
          .sticky-cell-container {
            background: var(--main-bg-color, white);
            border-bottom: dashed var(--primary-border, #cbcbcb) 1px;

            top: 0px;
            position: fixed;
            z-index: 300;
            width: 100%;
            max-height: 50%;

            padding-left: 10px;
            padding-right: 10px;
            padding-bottom: 10px;
            padding-top: 20px;

            overflow: auto;
          }

          .sticky-cell-container:empty {
            display: none;
          }

          .sticky-cell-container > :global(*) {
            margin: 20px;
          }
        `}</style>
      </div>
    );
  }
}

const PinnedPlaceHolderCell = () => (
  <div className="cell-placeholder">
    <span className="octicon">
      <LinkExternalOcticon />
    </span>
    <style jsx>{`
      .cell-placeholder {
        text-align: center;
        color: var(--main-fg-color);
        padding: 10px;
        opacity: var(--cell-placeholder-opacity, 0.3);
      }

      .octicon {
        transition: color 0.5s;
      }
    `}</style>
  </div>
);

export function getLanguageMode(metadata: ImmutableMap<*, *>): string {
  // First try codemirror_mode, then name, and fallback to 'text'
  const language = metadata.getIn(
    ["language_info", "codemirror_mode", "name"],
    metadata.getIn(
      ["language_info", "codemirror_mode"],
      metadata.getIn(["language_info", "name"], "text")
    )
  );
  return language;
}

const mapStateToProps = (state: Object) => ({
  theme: state.config.get("theme"),
  lastSaved: state.app.get("lastSaved"),
  kernelSpecDisplayName: state.app.get("kernelSpecDisplayName"),
  cellOrder: state.document.getIn(["notebook", "cellOrder"], ImmutableList()),
  cellMap: state.document.getIn(["notebook", "cellMap"], ImmutableMap()),
  transient: state.document.get("transient"),
  cellPagers: state.document.get("cellPagers"),
  cellFocused: state.document.get("cellFocused"),
  editorFocused: state.document.get("editorFocused"),
  stickyCells: state.document.get("stickyCells"),
  executionState: state.app.get("executionState"),
  models: state.comms.get("models"),
  language: getLanguageMode(
    state.document.getIn(["notebook", "metadata"], ImmutableMap())
  )
});

export class Notebook extends React.PureComponent<Props> {
  createCellElement: (s: string) => ?React$Element<any>;
  keyDown: (e: KeyboardEvent) => void;
  moveCell: (source: string, dest: string, above: boolean) => void;
  selectCell: (id: string) => void;
  renderCell: (id: string) => React$Element<any>;

  static defaultProps = {
    displayOrder,
    transforms,
    language: "python"
  };

  static contextTypes = {
    store: PropTypes.object
  };

  constructor(): void {
    super();
    this.createCellElement = this.createCellElement.bind(this);
    this.keyDown = this.keyDown.bind(this);
    this.moveCell = this.moveCell.bind(this);
    this.selectCell = this.selectCell.bind(this);
    this.renderCell = this.renderCell.bind(this);
  }

  componentDidMount(): void {
    document.addEventListener("keydown", this.keyDown);
  }

  componentWillUnmount(): void {
    document.removeEventListener("keydown", this.keyDown);
  }

  moveCell(sourceId: string, destinationId: string, above: boolean): void {
    this.context.store.dispatch(moveCell(sourceId, destinationId, above));
  }

  selectCell(id: string): void {
    this.context.store.dispatch(focusCell(id));
  }

  keyDown(e: KeyboardEvent): void {
    // If enter is not pressed, do nothing
    if (e.keyCode !== 13) {
      return;
    }

    let ctrlKeyPressed = e.ctrlKey;
    // Allow cmd + enter (macOS) to operate like ctrl + enter
    if (process.platform === "darwin") {
      ctrlKeyPressed = (e.metaKey || e.ctrlKey) && !(e.metaKey && e.ctrlKey);
    }

    const shiftXORctrl =
      (e.shiftKey || ctrlKeyPressed) && !(e.shiftKey && ctrlKeyPressed);
    if (!shiftXORctrl) {
      return;
    }

    if (!this.props.cellFocused) {
      return;
    }

    e.preventDefault();

    const cellMap = this.props.cellMap;
    const id = this.props.cellFocused;
    const cell = cellMap.get(id);

    if (e.shiftKey && !this.props.stickyCells.has(id)) {
      this.context.store.dispatch(focusNextCell(this.props.cellFocused, true));
      this.context.store.dispatch(focusNextCellEditor(id));
    }

    if (cell.get("cell_type") === "code") {
      this.context.store.dispatch(executeCell(id, cell.get("source")));
    }
  }

  renderStickyCells(): React.Node {
    const cellOrder = this.props.cellOrder;
    const stickyCells = cellOrder.filter(id => this.props.stickyCells.get(id));

    return (
      <StickyCellContainer>
        {stickyCells.map(id => (
          <div key={`sticky-cell-container-${id}`} className="sticky-cell">
            {this.renderCell(id)}
          </div>
        ))}
        <style jsx>{`
          .sticky-cell {
            padding-right: 20px;
          }
        `}</style>
      </StickyCellContainer>
    );
  }

  renderCell(id: string): React$Element<any> {
    const cell = this.props.cellMap.get(id);

    const running =
      this.props.transient.getIn(["cellMap", id, "status"]) === "busy";

    return (
      <Cell
        cell={cell}
        displayOrder={this.props.displayOrder}
        id={id}
        cellFocused={this.props.cellFocused}
        editorFocused={this.props.editorFocused}
        language={this.props.language}
        running={running}
        theme={this.props.theme}
        pagers={this.props.cellPagers.get(id)}
        transforms={this.props.transforms}
        models={this.props.models}
      />
    );
  }

  createCellElement(id: string): ?React$Element<any> {
    const isStickied = this.props.stickyCells.get(id);

    return (
      <div className="cell-container" key={`cell-container-${id}`}>
        {isStickied ? (
          <PinnedPlaceHolderCell />
        ) : (
          <DraggableCell
            moveCell={this.moveCell}
            id={id}
            selectCell={this.selectCell}
          >
            {this.renderCell(id)}
          </DraggableCell>
        )}
        <CellCreator key={`creator-${id}`} id={id} above={false} />
      </div>
    );
  }

  render(): ?React$Element<any> {
    return (
      <div>
        {/* Sticky cells */}
        {this.renderStickyCells()}
        {/* Actual cells! */}
        <div className="cells">
          <CellCreator id={this.props.cellOrder.get(0, null)} above />
          {this.props.cellOrder.map(this.createCellElement)}
        </div>
        <StatusBar
          lastSaved={this.props.lastSaved}
          kernelSpecDisplayName={this.props.kernelSpecDisplayName}
          executionState={this.props.executionState}
        />
        <style jsx>{`
          .cells {
            padding-top: 10px;
            padding-left: 10px;
            padding-right: 10px;
          }
        `}</style>
        <style jsx>{this.props.theme === "light" ? light : dark}</style>
      </div>
    );
  }
}

export const ConnectedNotebook = dragDropContext(HTML5Backend)(Notebook);
export default connect(mapStateToProps)(ConnectedNotebook);
