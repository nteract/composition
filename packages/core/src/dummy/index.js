// @flow
/* eslint-disable no-plusplus */

import * as Immutable from "immutable";
import { Subject } from "rxjs";
import {
  monocellNotebook,
  emptyCodeCell,
  appendCellToNotebook,
  emptyNotebook
} from "@nteract/commutable";
import { combineReducers, createStore } from "redux";

import { comms, config, core } from "../reducers";
import {
  makeNotebookContentRecord,
  makeRemoteKernelRecord,
  makeAppRecord,
  makeCommsRecord,
  makeKernelsRecord,
  makeDocumentRecord,
  makeContentsRecord,
  makeEntitiesRecord,
  makeStateRecord,
  createContentRef,
  createKernelRef
} from "../state";

export { dummyCommutable, dummy, dummyJSON } from "./dummy-nb";

const rootReducer = combineReducers({
  app: (state = makeAppRecord()) => state,
  comms,
  config,
  core
});

function hideCells(notebook) {
  return notebook.update("cellMap", cells =>
    notebook
      .get("cellOrder", Immutable.List())
      .reduce(
        (acc, id) => acc.setIn([id, "metadata", "inputHidden"], true),
        cells
      )
  );
}

/**
 * Creates a dummy notebook for Redux state for testing.
 *
 * @param {object} config - Configuration options for notebook
 * config.codeCellCount (number) - Number of empty code cells to be in notebook.
 * config.markdownCellCount (number) - Number of empty markdown cells to be in notebook.
 * config.hideAll (boolean) - Hide all cells in notebook
 * @returns {object} - A notebook for {@link DocumentRecord} for Redux store.
 * Created using the config object passed in.
 */
function buildDummyNotebook(config) {
  let notebook = monocellNotebook.setIn(
    ["metadata", "kernelspec", "name"],
    "python2"
  );

  if (config) {
    if (config.codeCellCount) {
      for (let i = 1; i < config.codeCellCount; i++) {
        notebook = appendCellToNotebook(notebook, emptyCodeCell);
      }
    }

    if (config.markdownCellCount) {
      for (let i = 0; i < config.markdownCellCount; i++) {
        notebook = appendCellToNotebook(
          notebook,
          emptyCodeCell.set("cell_type", "markdown")
        );
      }
    }

    if (config.hideAll) {
      notebook = hideCells(notebook);
    }
  }

  return notebook;
}

export function dummyStore(config: *) {
  const dummyNotebook = buildDummyNotebook(config);

  const frontendToShell = new Subject();
  const shellToFrontend = new Subject();
  const mockShell = Subject.create(frontendToShell, shellToFrontend);
  const channels = mockShell;

  const kernelRef = createKernelRef();
  const contentRef = createContentRef();

  // $FlowFixMe
  return createStore(rootReducer, {
    core: makeStateRecord({
      kernelRef,
      entities: makeEntitiesRecord({
        contents: makeContentsRecord({
          byRef: Immutable.Map({
            // $FlowFixMe: This really is a content ref, Flow can't handle typing it though
            [contentRef]: makeNotebookContentRecord({
              model: makeDocumentRecord({
                notebook: dummyNotebook,
                savedNotebook:
                  config && config.saved === true
                    ? dummyNotebook
                    : emptyNotebook,
                cellPagers: new Immutable.Map(),
                cellFocused:
                  config && config.codeCellCount > 1
                    ? dummyNotebook.get("cellOrder", Immutable.List()).get(1)
                    : null
              }),
              filepath:
                config && config.noFilename ? "" : "dummy-store-nb.ipynb"
            })
          })
        }),
        kernels: makeKernelsRecord({
          byRef: Immutable.Map({
            // $FlowFixMe: This really is a kernel ref, Flow can't handle typing it though
            [kernelRef]: makeRemoteKernelRecord({
              channels,
              status: "not connected"
            })
          })
        })
      })
    }),
    app: makeAppRecord({
      notificationSystem: {
        addNotification: () => {} // most of the time you'll want to mock this
      },
      githubToken: "TOKEN"
    }),
    config: Immutable.Map({
      theme: "light"
    }),
    comms: makeCommsRecord()
  });
}
