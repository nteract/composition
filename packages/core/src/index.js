//@flow

import * as actions from "./actions";
import * as constants from "./constants";
import * as middlewares from "./middlewares";
import * as records from "./records";
import * as reducers from "./reducers";
import * as components from "./components";

export { actions, constants, middlewares, records, reducers, components };
