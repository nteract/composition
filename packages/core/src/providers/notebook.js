/* eslint-disable no-return-assign */
/* @flow */
import * as React from "react";
import { DragDropContext as dragDropContext } from "react-dnd";
import HTML5Backend from "react-dnd-html5-backend";
import { connect } from "react-redux";
import {
  List as ImmutableList,
  Map as ImmutableMap,
  Set as ImmutableSet
} from "immutable";

import { displayOrder, transforms } from "@nteract/transforms";

import CellView from "../components/cell/cell";

import DraggableCell from "../components/draggable-cell";
import CellCreator from "./cell-creator";
import StatusBar from "../components/status-bar";

import {
  PinnedPlaceHolderCell,
  StickyCellContainer
} from "../components/pinned-cell";

import {
  focusCellEditor,
  focusPreviousCell,
  focusPreviousCellEditor,
  focusNextCell,
  focusNextCellEditor,
  moveCell,
  focusCell,
  executeCell
} from "../actions";

// NOTE: PropTypes are required for the sake of contextTypes
const PropTypes = require("prop-types");

const themes = require("../../themes");

type Props = {
  displayOrder: Array<string>,
  // TODO: Make types stricter, we have definitions _somewhere_
  cellOrder: ImmutableList<any>,
  cellMap: ImmutableMap<string, any>,
  transforms: Object,
  cellPagers: ImmutableMap<string, any>,
  stickyCells: ImmutableSet<string>,
  transient: ImmutableMap<string, any>,
  cellFocused: string,
  editorFocused: string,
  theme: string,
  lastSaved: Date,
  kernelSpecDisplayName: string,
  executionState: string,
  models: ImmutableMap<string, any>,
  language: string
};

export function getLanguageMode(metadata: ImmutableMap<*, *>): string {
  // First try codemirror_mode, then name, and fallback to 'text'
  const language = metadata.getIn(
    ["language_info", "codemirror_mode", "name"],
    metadata.getIn(
      ["language_info", "codemirror_mode"],
      metadata.getIn(["language_info", "name"], "text")
    )
  );
  return language;
}

const mapStateToProps = (state: Object) => ({
  theme: state.config.get("theme"),
  lastSaved: state.app.get("lastSaved"),
  kernelSpecDisplayName: state.app.get("kernelSpecDisplayName"),
  cellOrder: state.document.getIn(["notebook", "cellOrder"], ImmutableList()),
  cellMap: state.document.getIn(["notebook", "cellMap"], ImmutableMap()),
  transient: state.document.get("transient"),
  cellPagers: state.document.get("cellPagers"),
  cellFocused: state.document.get("cellFocused"),
  editorFocused: state.document.get("editorFocused"),
  stickyCells: state.document.get("stickyCells"),
  executionState: state.app.get("executionState"),
  models: state.comms.get("models"),
  language: getLanguageMode(
    state.document.getIn(["notebook", "metadata"], ImmutableMap())
  )
});

export class Notebook extends React.PureComponent<Props> {
  createCellElement: (s: string) => ?React$Element<any>;
  keyDown: (e: KeyboardEvent) => void;
  moveCell: (source: string, dest: string, above: boolean) => void;
  selectCell: (id: string) => void;
  renderCell: (id: string) => React$Element<any>;

  static defaultProps = {
    displayOrder,
    transforms,
    language: "python"
  };

  static contextTypes = {
    store: PropTypes.object
  };

  constructor(): void {
    super();
    this.createCellElement = this.createCellElement.bind(this);
    this.keyDown = this.keyDown.bind(this);
    this.moveCell = this.moveCell.bind(this);
    this.selectCell = this.selectCell.bind(this);
    this.renderCell = this.renderCell.bind(this);
  }

  componentDidMount(): void {
    document.addEventListener("keydown", this.keyDown);
  }

  componentWillUnmount(): void {
    document.removeEventListener("keydown", this.keyDown);
  }

  moveCell(sourceId: string, destinationId: string, above: boolean): void {
    this.context.store.dispatch(moveCell(sourceId, destinationId, above));
  }

  selectCell(id: string): void {
    this.context.store.dispatch(focusCell(id));
  }

  keyDown(e: KeyboardEvent): void {
    // If enter is not pressed, do nothing
    if (e.keyCode !== 13) {
      return;
    }

    let ctrlKeyPressed = e.ctrlKey;
    // Allow cmd + enter (macOS) to operate like ctrl + enter
    if (process.platform === "darwin") {
      ctrlKeyPressed = (e.metaKey || e.ctrlKey) && !(e.metaKey && e.ctrlKey);
    }

    const shiftXORctrl =
      (e.shiftKey || ctrlKeyPressed) && !(e.shiftKey && ctrlKeyPressed);
    if (!shiftXORctrl) {
      return;
    }

    if (!this.props.cellFocused) {
      return;
    }

    e.preventDefault();

    const cellMap = this.props.cellMap;
    const id = this.props.cellFocused;
    const cell = cellMap.get(id);

    if (e.shiftKey && !this.props.stickyCells.has(id)) {
      this.context.store.dispatch(focusNextCell(this.props.cellFocused, true));
      this.context.store.dispatch(focusNextCellEditor(id));
    }

    // $FlowFixMe: Cell might be undefined.
    if (cell.get("cell_type") === "code") {
      // $FlowFixMe: Cell might be undefined.
      this.context.store.dispatch(executeCell(id, cell.get("source")));
    }
  }

  renderStickyCells(): React.Node {
    const cellOrder = this.props.cellOrder;
    const stickyCells = cellOrder.filter(id => this.props.stickyCells.get(id));

    return (
      <StickyCellContainer>
        {stickyCells.map(id => (
          <div key={`sticky-cell-container-${id}`} className="sticky-cell">
            {this.renderCell(id)}
          </div>
        ))}
        <style jsx>{`
          .sticky-cell {
            padding-right: 20px;
          }
        `}</style>
      </StickyCellContainer>
    );
  }

  renderCell(id: string): React$Element<any> {
    const cell = this.props.cellMap.get(id);

    const running =
      this.props.transient.getIn(["cellMap", id, "status"]) === "busy";

    return (
      <CellView
        selectCell={() => {
          this.context.store.dispatch(focusCell(id));
        }}
        focusCellEditor={() => {
          this.context.store.dispatch(focusCellEditor(id));
        }}
        focusAboveCell={() => {
          this.context.store.dispatch(focusPreviousCell(id));
          this.context.store.dispatch(focusPreviousCellEditor(id));
        }}
        focusBelowCell={() => {
          this.context.store.dispatch(focusNextCell(id, true));
          this.context.store.dispatch(focusNextCellEditor(id));
        }}
        cell={cell}
        displayOrder={this.props.displayOrder}
        id={id}
        cellFocused={this.props.cellFocused}
        editorFocused={this.props.editorFocused}
        language={this.props.language}
        running={running}
        theme={this.props.theme}
        pagers={this.props.cellPagers.get(id)}
        transforms={this.props.transforms}
        models={this.props.models}
      />
    );
  }

  createCellElement(id: string): ?React$Element<any> {
    const isStickied = this.props.stickyCells.get(id);

    return (
      <div className="cell-container" key={`cell-container-${id}`}>
        {isStickied ? (
          <PinnedPlaceHolderCell />
        ) : (
          <DraggableCell
            moveCell={this.moveCell}
            id={id}
            selectCell={this.selectCell}
          >
            {this.renderCell(id)}
          </DraggableCell>
        )}
        <CellCreator key={`creator-${id}`} id={id} above={false} />
      </div>
    );
  }

  render(): ?React$Element<any> {
    return (
      <div>
        {/* Sticky cells */}
        {this.renderStickyCells()}
        {/* Actual cells! */}
        <div className="cells">
          <CellCreator id={this.props.cellOrder.get(0)} above />
          {this.props.cellOrder.map(this.createCellElement)}
        </div>
        <StatusBar
          lastSaved={this.props.lastSaved}
          kernelSpecDisplayName={this.props.kernelSpecDisplayName}
          executionState={this.props.executionState}
        />
        <style jsx>{`
          .cells {
            padding-top: 10px;
            padding-left: 10px;
            padding-right: 10px;
          }
        `}</style>
        <style>{`:root { ${themes[this.props.theme]} } `}</style>
      </div>
    );
  }
}

export const ConnectedNotebook = dragDropContext(HTML5Backend)(Notebook);
// $FlowFixMe: Flow can't figure out what to do with connect with one param.
export default connect(mapStateToProps)(ConnectedNotebook);
