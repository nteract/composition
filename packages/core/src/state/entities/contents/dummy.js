// @flow

import * as Immutable from "immutable";

export type EmptyModelRecordProps = {
  type: "unknown"
};

export const makeEmptyModel: Immutable.RecordFactory<
  EmptyModelRecordProps
> = Immutable.Record({
  type: "unknown"
});
export type EmptyModelRecord = Immutable.RecordOf<EmptyModelRecordProps>;

export type DummyContentRecordProps = {
  type: "dummy",
  assumedType: NotebookTypes,
  mimetype: ?string,
  lastSaved: null,
  filepath: string,
  model: EmptyModelRecord
};
export const makeDummyContentRecord: Immutable.RecordFactory<
  DummyContentRecordProps
> = Immutable.Record({
  type: "dummy",
  mimetype: null,
  assumedType: "unknown",
  lastSaved: null,
  filepath: "",
  model: makeEmptyModel()
});
export type DummyContentRecord = Immutable.RecordOf<DummyContentRecordProps>;
