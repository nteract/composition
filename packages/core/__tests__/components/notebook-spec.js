/* eslint-disable max-len */
import React from "react";
import Immutable from "immutable";
import { Provider } from "react-redux";
import { shallow, mount } from "enzyme";

import { displayOrder, transforms } from "@nteract/transforms-full";
import Cell from "../../src/components/cell/cell";
import { Notebook, getLanguageMode } from "../../src/components/notebook";

import { dummyStore, dummyCommutable } from "../../src/dummy";

const dummyCellStatuses = dummyCommutable
  .get("cellOrder")
  .reduce(
    (statuses, cellID) =>
      statuses.set(
        cellID,
        Immutable.fromJS({ outputHidden: false, inputHidden: false })
      ),
    new Immutable.Map()
  );

// Boilerplate test to make sure the testing setup is configured
describe("Notebook", () => {
  test("accepts an Immutable.List of cells", () => {
    const component = shallow(
      <Notebook
        notebook={dummyCommutable}
        transient={new Immutable.Map({ cellMap: new Immutable.Map() })}
        cellPagers={new Immutable.Map()}
        cellStatuses={new Immutable.Map()}
        stickyCells={new Immutable.Map()
          // Sticky the first cell of the notebook so that the sticky code gets
          // triggered.
          .set(dummyCommutable.getIn(["cellOrder", 0]), true)}
        CellComponent={Cell}
      />
    );
    expect(component).not.toBeNull();
  });
  test("implements the correct css spec", () => {
    const component = mount(
      <Provider store={dummyStore()}>
        <Notebook
          notebook={dummyCommutable}
          transient={
            new Immutable.Map({
              cellMap: new Immutable.Map()
            })
          }
          cellPagers={new Immutable.Map()}
          cellStatuses={dummyCellStatuses}
          stickyCells={new Immutable.Map()}
          displayOrder={displayOrder}
          transforms={transforms}
          CellComponent={Cell}
          models={new Immutable.Map({})}
        />
      </Provider>
    );
    expect(component.find(".notebook").length).toBeGreaterThan(0);
    expect(component.find(".notebook .cell").length).toBeGreaterThan(0);
    expect(component.find(".notebook .cell.text").length).toBeGreaterThan(0);
    expect(component.find(".notebook .cell.code").length).toBeGreaterThan(0);
    expect(component.find(".notebook .cell.unknown").length).toBe(0);
    expect(
      component.find(".notebook .cell.text .rendered").length
    ).toBeGreaterThan(0);
    expect(
      component.find(".notebook .cell.code .input-container").length
    ).toBeGreaterThan(0);
    expect(
      component.find(".notebook .cell.code .input-container .prompt").length
    ).toBeGreaterThan(0);
    expect(
      component.find(".notebook .cell.code .input-container .input").length
    ).toBeGreaterThan(0);
    expect(
      component.find(".notebook .cell.code .outputs").length
    ).toBeGreaterThan(0);
  });

  describe("getLanguageMode", () => {
    test("determines the right language from the notebook metadata", () => {
      const lang = getLanguageMode(dummyCommutable);
      expect(lang).toBe("ipython");

      const lang2 = getLanguageMode(
        dummyCommutable.setIn(
          ["metadata", "language_info", "codemirror_mode", "name"],
          "r"
        )
      );
      expect(lang2).toBe("r");
    });
  });

  describe("keyDown", () => {
    test("detects a cell execution keypress", () => {
      const focusedCell = dummyCommutable.getIn(["cellOrder", 1]);

      const context = { store: dummyStore() };

      context.store.dispatch = jest.fn();

      const component = shallow(
        <Notebook
          notebook={dummyCommutable}
          transient={new Immutable.Map({ cellMap: new Immutable.Map() })}
          cellPagers={new Immutable.Map()}
          cellStatuses={dummyCellStatuses}
          stickyCells={new Immutable.Map()}
          CellComponent={Cell}
          cellFocused={focusedCell}
        />,
        { context }
      );

      const inst = component.instance();

      const evt = new window.CustomEvent("keydown");
      evt.ctrlKey = true;
      evt.keyCode = 13;

      inst.keyDown(evt);

      expect(context.store.dispatch).toHaveBeenCalledWith({
        type: "EXECUTE_CELL",
        id: focusedCell,
        source: dummyCommutable.getIn(["cellMap", focusedCell, "source"])
      });
    });
    test("detects a focus to next cell keypress", () => {
      const focusedCell = dummyCommutable.getIn(["cellOrder", 1]);

      const context = { store: dummyStore() };

      context.store.dispatch = jest.fn();

      const component = shallow(
        <Notebook
          notebook={dummyCommutable}
          transient={new Immutable.Map({ cellMap: new Immutable.Map() })}
          cellPagers={new Immutable.Map()}
          cellStatuses={dummyCellStatuses}
          stickyCells={new Immutable.Map()}
          CellComponent={Cell}
          cellFocused={focusedCell}
        />,
        { context }
      );

      const inst = component.instance();

      const evt = new window.CustomEvent("keydown");
      evt.shiftKey = true;
      evt.keyCode = 13;

      inst.keyDown(evt);

      expect(context.store.dispatch).toHaveBeenCalledWith({
        type: "FOCUS_NEXT_CELL",
        id: focusedCell,
        createCellIfUndefined: true
      });
    });
    test("handles a focus to next cell keypress on a sticky cell", () => {
      const focusedCell = dummyCommutable.getIn(["cellOrder", 1]);

      const context = { store: dummyStore() };

      context.store.dispatch = jest.fn();

      const component = shallow(
        <Notebook
          notebook={dummyCommutable}
          transient={new Immutable.Map({ cellMap: new Immutable.Map() })}
          cellPagers={new Immutable.Map()}
          cellStatuses={dummyCellStatuses}
          stickyCells={new Immutable.Set([focusedCell])}
          CellComponent={Cell}
          cellFocused={focusedCell}
        />,
        { context }
      );

      const inst = component.instance();

      const evt = new window.CustomEvent("keydown");
      evt.shiftKey = true;
      evt.keyCode = 13;

      inst.keyDown(evt);

      expect(context.store.dispatch).not.toHaveBeenCalledWith({
        type: "FOCUS_NEXT_CELL",
        id: focusedCell,
        createCellIfUndefined: true
      });
    });
  });
});
