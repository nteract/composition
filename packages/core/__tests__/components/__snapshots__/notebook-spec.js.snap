// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sticky Cell Container creates a container when there are children 1`] = `
<div
  className="jsx-1455849675 jsx-1455849675"
>
  <div
    className="jsx-1455849675 jsx-1455849675 sticky-cells-placeholder"
  />
  <div
    className="jsx-1455849675 jsx-1455849675 sticky-cell-container"
  >
    <div>
      cell one
    </div>
    <div>
      cell two
    </div>
  </div>
</div>
`;

exports[`Sticky Cell Container returns null when there are no children 1`] = `null`;
